### PR TITLE
Added kubeval validation for output

### PR DESCRIPTION
**Pull request by surajnarwade**
This PR adds `kubeval` to vendor and use it as go library to validate generated kubernetes artifacts.

### CODE REVIEW

File: pkg/cmd/generate.go [hunk 1]
```diff
@@ -76,3 +79,27 @@ func Generate(paths []string) error {
 	}
 	return nil
 }
+
+func validateOutput(data []byte) {
+	results, err := kubeval.Validate(data, "output")
+	if err != nil {
+		errors.Wrap(err, "Unable to validate")
```
Comment (containscafeine) [inline pkg/cmd/generate.go]: This won't do anything, we need to either print or return this.

File: pkg/cmd/generate.go [hunk 2]
```diff
@@ -76,3 +79,27 @@ func Generate(paths []string) error {
...
+		os.Exit(1)
```
Comment (containscafeine) [inline pkg/cmd/generate.go]: Let's exit with -1 to maintain consistency in code

File: pkg/cmd/generate.go [hunk 3]
```diff
@@ -76,3 +79,27 @@ func Generate(paths []string) error {
...
+			fmt.Println("The document", result.FileName, "contains an invalid", result.Kind)
+			for _, desc := range result.Errors {
+				fmt.Println("--->", desc)
+			}
+		}
+
+	}
+
+	if !success {
+		os.Exit(1)
```
Comment (containscafeine) [inline pkg/cmd/generate.go]: Let's exit with -1 to maintain consistency in code

File: pkg/cmd/generate.go [hunk 4]
```diff
@@ -76,3 +79,27 @@ func Generate(paths []string) error {
...
+		errors.Wrap(err, "Unable to validate")
+		os.Exit(1)
+	}
+	success := true
+	for _, result := range results {
+		if len(result.Errors) > 0 {
+			success = false
+			fmt.Println("The document", result.FileName, "contains an invalid", result.Kind)
+			for _, desc := range result.Errors {
+				fmt.Println("--->", desc)
```
Comment (surajssd) [inline pkg/cmd/generate.go]: also let's return errors, than printing it!

File: pkg/cmd/generate.go (hunk 2 above)
Comment (surajssd) [inline pkg/cmd/generate.go]: better return it

File: pkg/cmd/generate.go [hunk 5]
```diff
@@ -76,3 +79,27 @@ func Generate(paths []string) error {
...
+	if err != nil {
+		errors.Wrap(err, "Unable to validate")
+		os.Exit(1)
+	}
+	success := true
+	for _, result := range results {
+		if len(result.Errors) > 0 {
+			success = false
+			fmt.Println("The document", result.FileName, "contains an invalid", result.Kind)
+			for _, desc := range result.Errors {
```
Comment (surajssd) [inline pkg/cmd/generate.go]: we don't even need to iterate over all the errors, just club them to form a single error and return it.

File: pkg/cmd/generate.go [hunk 6]
```diff
@@ -76,3 +79,27 @@ func Generate(paths []string) error {
 	}
 	return nil
 }
+
+func validateOutput(data []byte) {
```
Comment (containscafeine) [inline pkg/cmd/generate.go]: yep, let's return an error here and capture it wherever it is being called

File: pkg/cmd/generate.go (hunk 1 above)
Comment (kadel) [inline pkg/cmd/generate.go]: error should be returned

File: pkg/cmd/generate.go (hunk 3 above)
Comment (kadel) [inline pkg/cmd/generate.go]: exit shouldn't be called here. Let's just return an error.

**Comment by kadel** [comment]
@surajnarwade Have you tried this?

Because it's not working for me :-(

I've Kedge file with invalid name and invalid Service type and no errors were given.

```yaml
name: httpd
containers:
- image: centos/httpd
services:
  - name: httpd
    type: FOOBAR
    ports:
    - port: 8080
      targetPort: 80
```


I'm also afraid about if kubeval is reliable enough to use it like this :-(
We can't effort to have any false negatives with kubeval.

It might be better to have the command line option to enable this ( disabled by default for now.)

**Comment by kadel** [comment]
It might be better to use kubeval in unittest for now as described in https://github.com/kedgeproject/kedge/issues/184

**Comment by surajssd** [comment]
@kadel @containscafeine @surajnarwade this should also be part of `create`, `apply` and `delete` right? or just `generate`?

File: pkg/cmd/generate.go [hunk 7]
```diff
@@ -58,6 +59,8 @@ func Generate(paths []string) error {
 				return errors.Wrap(err, "failed to marshal object")
 			}
 
+			validateOutput(data)
```
Comment (kadel) [inline pkg/cmd/generate.go]: validation is done only for `generate`?
Any reason why not doing it also for `apply` and `create`?

Comment (surajssd) [inline pkg/cmd/generate.go]: wow timing! :rofl:

**Comment by surajnarwade** [comment]
@kadel @surajssd , yeah it should part of `create`, `apply` and `delete` as well, according to @kadel , shall I add additional flag to test or something like,
```
kedge validate -f foo.yml
```
as experimental thing only

**Comment by kadel** [comment]
> @kadel @surajssd , yeah it should part of create, apply and delete as well, according to @kadel , shall I add additional flag to test or something like,
> 
> kedge validate -f foo.yml
> as experimental thing only

not like this, as this looks like it's validating Kedge file.



how about this:
```
kedge generate -f foo.yml --validate-kubernetes=true
```

if this proves to be useful, we can just change the default value and have it enabled by default.
If someone will need to disable it will be possible to do `--validate-kubernetes=false`

**Comment by kadel** [comment]
Does someone actually tested how reliable Kubeval is?

**Comment by kadel** [comment]
It looks quite useless, if it tells that something like this is valid

```
---
apiVersion: extensions/v1beta1
kind: Deployment
metadata:
  creationTimestamp: null
  labels:
    app: httpd
  name: httpd
spec:
  strategy: {}
  template:
    metadata:
      creationTimestamp: null
      labels:
        app: httpd
      name: httpd
    spec:
      containers:
      - image: centos/httpd
        name: httpd
        resources: {}
status: {}
---
apiVersion: v1
kind: Service
metadata:
  creationTimestamp: null
  labels:
    app: httpd
  name: INVALID-e_f
spec:
  ports:
  - port: 8080
    targetPort: 80
  selector:
    app: httpd
  type: INVALID
status:
  loadBalancer: {}
```

```
▶ kubeval f.yml                                           
The document  contains a valid 
The document f.yml contains a valid Deployment
The document f.yml contains a valid Service

```

**Comment by surajnarwade** [comment]
@kadel , I have filed issue on kubeval which can be tracked here https://github.com/garethr/kubeval/issues/18

**Comment by cdrage** [comment]
Correct me if I'm wrong @surajnarwade but this is still blocked on upstream-code right?

**Comment by surajssd** [comment]
Yes this is blocked on the upstream code but we have decided that we go ahead with the the spec that we have by adding the changes to the spec manually as we need!

**Comment by cdrage** [comment]
@surajssd So if we're adding the changes manually, this PR can be closed?

**Comment by surajssd** [comment]
Not sure, how is @surajnarwade taking this forward either with this PR or a new one!

**Comment by surajnarwade** [comment]
@surajssd @cdrage , few changes needed, since upstream is fixed now, I will send a fix soon

**Comment by surajnarwade** [comment]
@kadel @surajssd @cdrage can you please try this PR ?

File: glide.lock [hunk 8]
```diff
@@ -1,10 +1,12 @@
-hash: f3e767e8346d32c0a86b9c441a9f5a580a4da1e4e9b201cffc7e2b313149bc5d
-updated: 2017-11-14T15:40:02.556527845+05:30
+hash: 9be3af04ee1417f46551109c7ed120aa3bb93a56a471144a7be35464d180f775
```
Comment (surajssd) [inline glide.lock]: The vendor changes, glide changes should go in one commit and the rest of the code changes should go in another commit!

File: glide.lock [hunk 9]
```diff
@@ -1,10 +1,12 @@
...
+updated: 2017-11-22T14:55:30.531845973+05:30
 imports:
 - name: github.com/davecgh/go-spew
   version: 5215b55f46b2b919f50a1df0eaa5886afe4e3b3d
   subpackages:
   - spew
+- name: github.com/fsnotify/fsnotify
```
Comment (surajssd) [inline glide.lock]: about the vendor update how did you make the changes? Because for the kubeval repo I also see the README.md and all irrelevant files in there?

File: pkg/cmd/util.go [hunk 10]
```diff
@@ -153,3 +154,28 @@ func replaceWithEnv(in []byte) []byte {
...
+			errs = append(errs, err)
+		}
+	}
+	if len(errs) > 0 {
+		return fmt.Errorf("%s", strings.Join(errs, "\n"))
+	}
+
+	return nil
+
+}
```
Comment (surajssd) [inline pkg/cmd/util.go]: can you add some tests to see what kind of output will be the wrong output?

**Comment by surajnarwade** [comment]
Since kubeval is not maintained, I am closing this PR
